### DISCUSSION TITLE
External RTC Module Integration (i.e. DS3231)

### DISCUSSION THREAD

**Discussion by biswajitkgp** [tobozo/WiFiManagerTz#2] [Q&A]
Hi tobozo!

First of all congratulations for making such an awesome library! It is really helpful. Can you help me integrating a DS3231 with the ESP32?

Whenever the ESP32 syncs time with the NTP Server, it should also set the time of the DS3231 Module. You can like add the code for writing time to the DS3231 maybe in the following function in your NTP.cpp file

// Callback function (get's called when time adjusts via NTP)
  void timeavailable(struct timeval *t)
  {
    Serial.println("Got time adjustment from NTP!");
    struct tm timeInfo;
    getLocalTime(&timeInfo, 1000);
    Serial.println(&timeInfo, "%A, %B %d %Y %H:%M:%S zone %Z %z ");
  }

Any help will be really appreciated. Thanks in advance.

**Comment by tobozo** [comment]
hey @biswajitkgp thanks for your feedback :+1: 

I'll add a callback setter for this and push changes soon, just found a few minor bugs I need to fix before :wink:

**Reply by biswajitkgp** [reply]
Thank you so much for considering this update. In the meantime, can you help me with the following?

![Screenshot 2023-01-17 213752](https://user-images.githubusercontent.com/121790066/212951095-00def7de-e330-4d53-8d1c-61a797a21306.png)
Suppose I want to change the Title of the page from **WiFiManager** to **Clock Manager**, can you help me where to make the changes?

Also I do not want to display **ESP32_9E12CFA4**. So please help me from where I can edit these. Thanks.

**Reply by tobozo** [reply]
kinda out of scope here, maybe ask on the WiFiManager discussion page?

**Reply by biswajitkgp** [reply]
Okay i will ask on WiFiManager discussion page. Thanks for your response.

**Comment by tobozo** [comment]
just pushed some changes on [1.3.2 branch](https://github.com/tobozo/WiFiManagerTz/tree/1.3.2)

A custom `void(struct timeval *t)` callback can now be attached by adding this statement **before `WiFiManagerNS::init()`**:

```cpp
WiFiManagerNS::NTP::onTimeAvailable( &your_custom_callback );
```

Just implement your RTC.adjust in `void your_custom_callback( struct timeval *t)` and you're done.

Please let me know how this worked for you so I can prepare a new release.

**Reply by biswajitkgp** [reply]
Let me try this out and inform.

**Reply by biswajitkgp** [reply]
@tobozo The callback works absolutely fine! Kudos to you!
